tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/annotations/AnnotationHelper.java

```diff
@@ -0,0 +1,156 @@
+
+    /**
+     * Annotation as implemented on passed in type or parent of that type, works for both super classes and interfaces.
+     *
+     * @param clazz - Class type to scan for annotations
+     * @param annotation - Annotation type to scan for
+     * @return - First matching annotation found in depth first search, or null if not found
+     */
+    public static <T extends Annotation> T getSuperTypeAnnotation(Class<?> clazz, Class<T> annotation) {
+        if (clazz.isAnnotationPresent(annotation)) {
+            return clazz.getAnnotation(annotation);
+        }
+
+        for (Class<?> ifaces : getParentClasses(clazz)) {
+            T superAnnotation = getSuperTypeAnnotation(ifaces, annotation);
+            if (superAnnotation != null) {
+                return superAnnotation;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Depth first search up the Type hierarchy to find a matching annotation,  Types which do not implement the
+     * specified method signature are ignored.
+     *
+     * @param annotation - Annotation type to scan for
+     * @param clazz - Class type to scan for matching annotations
+     * @param methodSignature - Method to search annotation for
+     * @return - First found matching annotation or null
+     */
+    public static <T extends Annotation> T getMethodAnnotation(
+            Class<T> annotation, Class<?> clazz, MethodSignature methodSignature) {
+
+        Method method;
+        try {
+            method = clazz.getMethod(methodSignature.getMethodName(), methodSignature.getParameterTypes());
+        } catch (NoSuchMethodException e) {
+            return null;
+        }
+
+        if (method.isAnnotationPresent(annotation)) {
+            return method.getAnnotation(annotation);
+        }
+
+        for (Class<?> iface : getParentClasses(clazz)) {
+            T foundAnnotation = getMethodAnnotation(annotation, iface, methodSignature);
+            if (foundAnnotation != null) {
+                return foundAnnotation;
+            }
+        }
+
+        return null;
+    }
+
+    private static List<Class<?>> getParentClasses(Class<?> clazz) {
+        ImmutableList.Builder<Class<?>> builder = ImmutableList.builder();
+        if (clazz.getSuperclass() != null) {
+            builder.add(clazz.getSuperclass());
+        }
+
+        return builder.addAll(Arrays.asList(clazz.getInterfaces())).build();
+    }
+
+    public static final class MethodSignature {
+        private final String methodName;
+        private final Class<?>[] parameterTypes;
+
+        private MethodSignature(String methodName, Class<?>... parameterTypes) {
+            this.methodName = checkNotNull(methodName);
+            this.parameterTypes = parameterTypes != null ? parameterTypes : new Class<?>[]{};
+        }
+
+        public String getMethodName() {
+            return methodName;
+        }
+
+        public Class<?>[] getParameterTypes() {
+            return parameterTypes;
```

Comment:
I think we can get away without exposing the getters, and probably move this to a top level class.

```
public final class MethodSignature {

    private final String name;
    private final List<Class<?>> parameterTypes;

    private MethodSignature(String name, List<Class<?>> parameterTypes) {
        this.name = name;
        this.parameterTypes = ImmutableList.copyOf(parameterTypes);
    }

    public static MethodSignature from(Method method) {
        return new MethodSignature(method.getName(), ImmutableList.copyOf(method.getParameterTypes()));
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof MethodSignature)) {
            return false;
        }

        MethodSignature that = (MethodSignature) obj;
        return name.equals(that.name) && parameterTypes.equals(that.parameterTypes);
    }

    @Override
    public int hashCode() {
        return 31 * name.hashCode() + parameterTypes.hashCode();
    }

    @Override
    public String toString() {
        return "MethodSignature{"
                + "methodName='" + name + '\''
                + ", parameterTypes=" + parameterTypes
                + '}';
    }

}
```